### ISSUE TITLE
type cast double to int in amici_interface_matlab.cpp

### ISSUE THREAD

**Issue by thomassligon** [ICB-DCM/AMICI#228]
Compiler error:
C:\data\Tom\Research\AMICI\src\amici_interface_matlab.cpp(140): error C2440: 'type cast': cannot convert from 'double' to 'amici::LinearMultistepMethod'
The error occurs first in line 140, but a similar error also occurs in lines 141-150.
The cause for the error is in the definition of readOptionScalar, in the statement in line 57:
`(TYPE)mxGetScalar(mxGetProperty(prhs[RHS_OPTIONS], 0, #OPTION));`
This statement is making a type cast from double to int.
mxGetScalar returns double:
https://de.mathworks.com/help/matlab/apiref/mxgetscalar.html
(TYPE) is defined in AMICI as an enum, which is a type of int.
double is double precision floating point type, usually IEEE-754 64 bit floating point type.
This is a static type cast and amounts to treating the floating-point data as if it were integer, which can go wrong depending on the hardware architecture. This is why it is an error and should be replaced by a dynamic type cast, which will actually do conversion, including rounding, if necessary.
New line 57:
`            (TYPE)int(mxGetScalar(mxGetProperty(prhs[RHS_OPTIONS], 0, #OPTION)));   \`

**Comment by dweindl** [comment]
@thomassligon Which compiler are you using?

**Comment by thomassligon** [comment]
I am using Visual Studio 2017 C++.

**Comment by FFroehlich** [comment]
We officially do not support MSVS products due to their notorious lack of compliance with C and C++ standards. Amiwrap should produce an according warning in line 18.
Could you please tell me the output of `mex.getCompilerConfigurations('c++').Name` on your systems?

Nevertheless the issue is indeed problematic and we are aware of it (https://github.com/ICB-DCM/AMICI/issues/181).

**Comment by thomassligon** [comment]
Hello Daniel,
the remark "notorious lack of compliance" is difficult for me to understand. Could you be a bit more specific?
Here is the output of amiwrap and getCompilerConfigurations:
```
>> amiwrap(modelAMICI,modelAMICI,pathName,false);
Generating model struct ...
Parsing model struct ...

Generating C code ...
headers | wrapfunctions | CMakeLists | main | 
Compiling mex file ...
wrapfunctions | Building with 'Microsoft Visual C++ 2017'.
MEX completed successfully.
linking | Building with 'Microsoft Visual C++ 2017 (C)'.
MEX completed successfully.
Generating M code ...
>> mex.getCompilerConfigurations('c++').Name

ans =

    'Microsoft Visual C++ 2017'

```

**Comment by FFroehlich** [comment]
It took Microsoft 14 years to support C99 features like mixing declarations with code (see https://stackoverflow.com/questions/9610747/which-c99-features-are-available-in-the-ms-visual-studio-compiler) and after all, full C99 support is likely to never happen (https://herbsutter.com/2012/05/03/reader-qa-what-about-vc-and-c99/).

We continuously encountered problems when trying to ensure compatibility with MSVS, in particular with versions prior to 2013 (people often use old MSVS version because their code does not compile on newer versions ...) where standard support is generally poor https://msdn.microsoft.com/en-us/library/hh567368.aspx

With MSVS 2017 things might have improved, but many developers, including those of SUNDIALS and SuiteSparse have decided to not support MSVS. This makes it extremely painful to ensure compatibility as there are no automated build scripts that could be used for CI and respective testing.

**Comment by thomassligon** [comment]
Fabian,
thanks for the good discussion. I also started a discussion here:
https://social.msdn.microsoft.com/Forums/vstudio/en-US/1343bd04-685e-4e45-8801-65fb22860082/error-c2440-type-cast-from-double-to-int?forum=vcgeneral

Here is an excerpt from it, confirming your casting of double directly to enum:

 ( http://eel.is/c++draft/expr.static.cast#10 ):
[expr.static.cast]/10 ... A value of floating-point type can also be explicitly converted to an enumeration type. The resulting value is the same as converting the original value to the underlying type of the enumeration (4.9), and subsequently to the enumeration type.

http://en.cppreference.com/w/cpp/language/static_cast
"8) ...If you read a bit further down, you'll find this sentence in the very same paragraph:
"A value of a floating-point type can also be converted to any complete enumeration type"

In summary, it appears that the error message C2440 is not justified, and my suggestion should not be necessary, but should work on all compilers.
With this small change, Visual Studio 2017 works fine for AMICI and Sundials, at least as far as I have compiled them. In my experience, 2012 and 2013 are ancient history in the IT timeframe. For this specific issue, I will look for a way to submit a bug report to Microsoft.

**Comment by FFroehlich** [comment]
But please be aware that 
changing lin 57 to 
`(TYPE)int(mxGetScalar(mxGetProperty(prhs[RHS_OPTIONS], 0, #OPTION))); \` 
will probably lead to unexpected results as atol and rtol, which are in fact doubles are also read in through this function. As these values are typically <<1, the double and int cast are likely to set them to 0.

Regarding the dynamic casts: I do not understand why a dynamic cast should help in this case. The static cast should simply truncate (which is evidently bad if an integer value 1 is stored as float as 0.9999999999....). Are `int`/`double` really polymorphic? What does the dynamic cast actually do? How can it safely do the right thing for 1.5? I did not find a good reference for these questsions :(

Judging by the discussion you linked, throwing an error is in this case is appropriate under the C++/CX (and another good example of noncompliance with standards) which is an extension of C++  and is likely to be only supported by MSVS. Again, what I do not understand is how the dynamic cast helps. If the float will be cast to an out of range int that leads to undefined behaviour, but that will still be the case with a dynamic cast in between.

**Comment by thomassligon** [comment]
After this long discussion, I think a dynamic cast is a bad idea. Also, my suggestion for line 57 is wrong, because I was assuming that this is always an enum, which is always an int, but atol and rtol actually require doubles. That leaves me with no easy solution for Visual Studio. This needs to do conversion dependent on (TYPE). For the case of VS, I want to contact support and see, for example,  if this is controlled by a compiler option.

**Comment by dweindl** [comment]
Does changing `(TYPE)mxGetScalar...` to `static_cast<TYPE>(mxGetScalar...)` get rid of the warning?

If not, we can turn the `readOptionScalar` macro into `double readOptionScalar(const char* option)` and use `udata.lmm = static_cast<LinearMultistepMethod>(static_cast<int>readOptionScalar(lmm)));`

The `goto freturn` should be changed to `throw`.

**Comment by thomassligon** [comment]
No, using the static_cast just slightly changes the error message
error C2440: 'static_cast': cannot convert from 'double' to 'amici::LinearMultistepMethod'
PS I just submitted a bug report, but I don't expect miracles overnight.

**Comment by thomassligon** [comment]
I submitted a bug report to Microsoft and its status has changed first to "triaged" and then to "under consideration".
I also found some documentation here
https://msdn.microsoft.com/en-us/library/mt490614.aspx
I have added
/std:c++latest
to COMPFLAGS in C:\Users\thoma\AppData\Roaming\MathWorks\MATLAB\R2017b\mex_C_win64.xml
but it didn’t help.

**Comment by FFroehlich** [comment]
I just checked and we actually specify the CXXFLAGS to include `std=c++0x` (compileAndLinkMode` line 41). Thus the compiler should not throw any errors/warnings related to c++17.

**Comment by thomassligon** [comment]
This is now working fine for me. Thanks.
```
Generating C code ...
headers | wrapfunctions | CMakeLists | main | 
Compiling mex file ...
AMICI base files | Building with 'Microsoft Visual C++ 2017'.
MEX completed successfully.
ffuns | Building with 'Microsoft Visual C++ 2017'.
MEX completed successfully.
wrapfunctions | Building with 'Microsoft Visual C++ 2017'.
MEX completed successfully.
linking | Building with 'Microsoft Visual C++ 2017 (C)'.
MEX completed successfully.
Generating M code ...
```
